### ISSUE TITLE
S3 snapshot failure - files not found

### ISSUE THREAD

**Issue by raags** [elastic/elasticsearch#18632]
**Elasticsearch version**: 1.7.5

**JVM version**: 
```
java version "1.7.0_76"
Java(TM) SE Runtime Environment (build 1.7.0_76-b13)
Java HotSpot(TM) 64-Bit Server VM (build 24.76-b04, mixed mode)
```

**OS version**:
```
Distributor ID: Ubuntu
Description:    Ubuntu 12.04.5 LTS
Release:        12.04
Codename:       precise
```
**Description of the problem including expected versus actual behavior**:

I'm using elasticsearch-cloud-aws/2.7.1 to snapshot to s3 storage and restore to a backup cluster. This works as expected, but on some occasions the restore fails with the following exception :

```
[2016-05-30 01:30:09,837][WARN ][cluster.action.shard     ] [Node-02] [index_blah][4] received shard failed for [index_blah][4], node[fWO2LYbiTyaNGNjKWx0DXQ], [P], restoring[prod:snapshot_160530010001], s[INITIALIZING], unassigned_info[[reason=ALLOCATION_FAILED], at[2016-05-30T01:30:09.722Z], details[shard failure [failed recovery][IndexShardGatewayRecoveryException[[index_blah][4] failed recovery]; nested: IndexShardRestoreFailedException[[index_blah][4] restore failed]; nested: IndexShardRestoreFailedException[[index_blah][4] failed to restore snapshot [snapshot_160530010001]]; nested: IndexShardRestoreFailedException[[index_blah][4] Failed to recover index]; nested: FileNotFoundException[Blob object [__idw] not found: The specified key does not exist. (Service: Amazon S3; Status Code: 404; Error Code: NoSuchKey; Request ID: 660EA6BE976AF53A)]; ]]], indexUUID [KILcs5zETQmcm-zhsoLymQ], reason [shard failure [failed recovery][IndexShardGatewayRecoveryException[[index_blah][4] failed recovery]; nested: IndexShardRestoreFailedException[[index_blah][4] restore failed]; nested: IndexShardRestoreFailedException[[index_blah][4] failed to restore snapshot [snapshot_160530010001]]; nested: IndexShardRestoreFailedException[[index_blah][4] Failed to recover index]; nested: FileNotFoundException[Blob object [__idw] not found: The specified key does not exist. (Service: Amazon S3; Status Code: 404; Error Code: NoSuchKey; Request ID: 16AA079D6DCE946A)]; ]]
```

This seems to indicate that the snapshot did not complete successfully, but on the production cluster I can see the log line which indicates success :

```
[2016-05-30 01:01:28,009][INFO ][snapshots                ] [prod-03] snapshot [prod:snapshot_160530010001] is done
```

It seems like the snapshot is not completely transferred, or a could be a network failure. We also have another cluster in azure which works fine, so seems like an S3 specific issue.

**Comment by dadoonet** [comment]
Thanks for reporting this.
Could you check manually if you can see the file on S3 and what are its properties?

If you run again the same restore command does it fail exactly the same way?

Cc @imotov

**Comment by raags** [comment]
Yes, I did. I don't see the files on S3; and yes it fails for that particular snapshot. Its basically a corrupt snapshot. It fails on the same file.

**Comment by dadoonet** [comment]
Thanks @raags.

The snapshot operation should have raised an exception in that case. 
So I can see only 2 explanations:

* Bug on AWS/S3 side: could be a bug in their SDK also which may be swallow an exception while running the snapshot.
* Bug in AWS cloud plugin for elasticsearch which would swallow the exception while running the snapshot. I read again the code in https://github.com/elastic/elasticsearch-cloud-aws/blob/v2.7.1/src/main/java/org/elasticsearch/cloud/aws/blobstore/DefaultS3OutputStream.java and did not see anything obvious.

We are on the way of upgrading the SDK to 1.10.69 with this PR: https://github.com/elastic/elasticsearch-cloud-aws/pull/273 which might hopefully fix this behavior...

Will be also merged for 2.x series with #18048 and for 5.0 with #17784.

I'm unsure on how we can reproduce such an error though. Apart what we already do to simulate write or read errors: https://github.com/elastic/elasticsearch-cloud-aws/blob/v2.7.1/src/test/java/org/elasticsearch/cloud/aws/TestAmazonS3.java#L75

I'm leaving it open until @imotov comments on it as well.

**Comment by imotov** [comment]
@dadoonet the restore service is trying to restore these files, which means that snapshot service thought that these files were successfully written to S3. But it looks like they either were not written or they somehow disappeared afterwards. 

@raags was it a one time issue or you can reproduce it in your setup?

**Comment by raags** [comment]
@imotov we have now correlated this occurrence with the snapshot clean up job that runs once a day. The snapshot that fails to restore happens during the cleanup. In cleanup we catch the ConcurrentExecution exception and wait/retry until the snapshot completes. Could this be causing removal of some snapshot files? Also, oddly this doesn't happen on other clusters that have the same setup (snapshot/cleanup schedules)

**Comment by imotov** [comment]
@raags It looks like there is a tricky race condition in snapshot deletion. If a delete operation starts while a snapshot is being created, the delete operation indeed fails with ConcurrentSnapshotExecutionException. However, when a snapshot operation starts while a delete operation is in progress, the snapshot operation doesn't check for a running delete operation and as a result the delete operation and the snapshot operation can right simultaneously as long as a delete operation starts first . There are few more conditions that should take place sequence-wise for this condition to occur, but I can see how some files used by an in-progress snapshot might be deleted during a delete operation and cause the error that you've posted. I would guess, it might not be happening on other clusters because their cleanup and snapshot schedules are aligned slightly differently or because delete operations are slightly faster on these clusters for whatever reason and manage to finish before the next snapshot starts.

@abeyad I know you are collecting possible race conditions for snapshot/restore operations. Did we capture this condition somewhere else already or should we use this issue to track it?

**Comment by abeyad** [comment]
Yes, this is a race condition that we encountered amongst others in the past.  However, we should keep this issue open as I hadn't gotten around to capturing this one yet.  Another issue that also presents race conditions is found here: https://github.com/elastic/elasticsearch/issues/18551

Though its not the same, it relates in the sense that it reveals race conditions as a result of a separation between using cluster state and the repository to maintain snapshots without added synchronization around constructs around those. 

Another simple example is if we call the get snapshots API, first we look in the cluster state, get the current snapshot, then look in the repository for the rest of the snapshots. In between looking in the cluster state for running snapshots and looking in the repository, that running snapshot could have completed and gotten stored in the repository, so it would be counted twice.  This is a benign example but is a manifestation of this class of race conditions. 

We have discussed these and been capturing them along the way and are in the process of formulating ideas to fix them.

**Comment by raags** [comment]
Thanks for the quick turn around. The once a day clean up task used to take a long time, which is why the snapshot conflict happened. We have changed the cleanup interval to every other hour now, so its faster.

**Comment by abeyad** [comment]
The issue with restores mentioned in this issue will largely go away with the following fix: https://github.com/elastic/elasticsearch/pull/19853.  In this PR, we check if a restore is in progress before attempting to execute a snapshot deletion.

There is still a slight possibility for race conditions, which is a larger issue that needs to be tackled and I opened an issue for it: https://github.com/elastic/elasticsearch/issues/19957

Closing this issue.
